### DISCUSSION TITLE
Invalid sudoers file

### DISCUSSION THREAD

**Discussion by CyberPingU** [gravitational/teleport#35474] [Q&A]
Adding this to a role file

```
    host_groups: [wheel]
    host_sudoers: [
       "%wheel=(ALL) NOPASSWD: ALL" 
    ]

```
leads to

```
Dec  7 09:21:03 osiris teleport[1482238]: 2023-12-07T09:21:03Z             WARN "parse error in stdin near line 1\n\n\tvisudo: invalid sudoers file" regular/sshserver.go:1743
Dec  7 09:21:04 osiris teleport[1482238]: 2023-12-07T09:21:04Z             ERRO failure handling SSH "shell" request error:[
Dec  7 09:21:04 osiris teleport[1482238]: ERROR REPORT:
Dec  7 09:21:04 osiris teleport[1482238]: Original Error: *errors.errorString visudo: invalid sudoers file
Dec  7 09:21:04 osiris teleport[1482238]: Stack Trace:
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/utils/host/hostusers.go:155 github.com/gravitational/teleport/lib/utils/host.CheckSudoers
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/usermgmt_linux.go:114 github.com/gravitational/teleport/lib/srv.(*HostSudoersProvisioningBackend).CheckSudoers
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/usermgmt_linux.go:141 github.com/gravitational/teleport/lib/srv.(*HostSudoersProvisioningBackend).WriteSudoersFile
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/usermgmt.go:205 github.com/gravitational/teleport/lib/srv.(*HostSudoersManagement).WriteSudoers
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/sess.go:248 github.com/gravitational/teleport/lib/srv.(*SessionRegistry).TryWriteSudoersFile
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1709 github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1577 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests
Dec  7 09:21:04 osiris teleport[1482238]: #011github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1310 github.com/gravitational/teleport/lib/srv/regular.(*Server).HandleNewChan.func1
Dec  7 09:21:04 osiris teleport[1482238]: #011runtime/asm_amd64.s:1650 runtime.goexit
Dec  7 09:21:04 osiris teleport[1482238]: User Message: parse error in stdin near line 1
Dec  7 09:21:04 osiris teleport[1482238]: #011visudo: invalid sudoers file] regular/sshserver.go:2094
```
If I replace this with

```
    host_groups: [wheel]
    host_sudoers: [
       "ALL=(ALL) NOPASSWD: ALL" 
    ]
```

Everything works... am I wrong somewhere?

**Comment by webvictim** [comment]
The syntax of your original sudoers file is indeed incorrect.

I think you actually want this:
```yaml
    host_groups: [wheel]
    host_sudoers: [
       "%wheel ALL=(ALL) NOPASSWD: ALL" 
    ]
```

This grants any user in the wheel group (`%wheel`) the ability to run any command with no password (`NOPASSWD: ALL`) as any user (`ALL`) on any host (`(ALL)`)

**Comment by CyberPingU** [comment]
Hi webvictim,
to be honest I tried some combinations of those words, I might have copypasted just the wrong  one...
again with:

```
    host_groups: [sudo]
    host_sudoers: [
      "%sudo  ALL=(ALL) NOPASSWD: ALL"
    ] 
```
I get
 
```
2023-12-07T23:28:36.622021+01:00 vega teleport[465159]: 2023-12-07T23:28:36+01:00 WARN             "\n\tvisudo: invalid sudoers file" regular/sshserver.go:1743
2023-12-07T23:28:36.681125+01:00 vega teleport[465159]: 2023-12-07T23:28:36+01:00 ERRO             failure handling SSH "shell" request error:[
2023-12-07T23:28:36.681238+01:00 vega teleport[465159]: ERROR REPORT:
2023-12-07T23:28:36.681265+01:00 vega teleport[465159]: Original Error: *errors.errorString visudo: invalid sudoers file
2023-12-07T23:28:36.681287+01:00 vega teleport[465159]: Stack Trace:
2023-12-07T23:28:36.681309+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/utils/host/hostusers.go:155 github.com/gravitational/teleport/lib/utils/host.CheckSudoers
2023-12-07T23:28:36.681331+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/usermgmt_linux.go:114 github.com/gravitational/teleport/lib/srv.(*HostSudoersProvisioningBackend).CheckSudoers
2023-12-07T23:28:36.681354+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/usermgmt_linux.go:141 github.com/gravitational/teleport/lib/srv.(*HostSudoersProvisioningBackend).WriteSudoersFile
2023-12-07T23:28:36.681377+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/usermgmt.go:205 github.com/gravitational/teleport/lib/srv.(*HostSudoersManagement).WriteSudoers
2023-12-07T23:28:36.681398+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/sess.go:248 github.com/gravitational/teleport/lib/srv.(*SessionRegistry).TryWriteSudoersFile
2023-12-07T23:28:36.681422+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1709 github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch
2023-12-07T23:28:36.681446+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1577 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests
2023-12-07T23:28:36.681468+01:00 vega teleport[465159]: #011github.com/gravitational/teleport/lib/srv/r
```

Is there anything about tabs or spaces I should care of in writing the line?

I even copy-pasted your code, but it doesn't work. 
There's no other "wheel" rule-group in  the  system's sudoers' that might conflict.
By adding the line "by hand"
`      "%wheel  ALL=(ALL) NOPASSWD: ALL"`
 with visudo, and running a syntax check with visudo -c, I get no errors.
Same with
`"%wheel ALL=(ALL:ALL) NOPASSWD: ALL"`

**Comment by code-dave** [comment]
I am getting this same error:

`Dec 20 20:04:35 __REDACTED__ teleport[17226]: 2023-12-20T20:04:35Z             WARN "parse error in stdin near line 2\n\n\tvisudo: invalid sudoers file" regular/sshserver.go:1743
Dec 20 20:04:35 __REDACTED__ teleport[17226]: 2023-12-20T20:04:35Z             ERRO failure handling SSH "shell" request error:[
Dec 20 20:04:35 __REDACTED__ teleport[17226]: ERROR REPORT:
Dec 20 20:04:35 __REDACTED__ teleport[17226]: Original Error: *errors.errorString visudo: invalid sudoers file
Dec 20 20:04:35 __REDACTED__teleport[17226]: Stack Trace:
Dec 20 20:04:35 __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/utils/host/hostusers.go:155 github.com/gravitational/teleport/lib/utils/host.CheckSudoers
Dec 20 20:04:35 __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/usermgmt_linux.go:114 github.com/gravitational/teleport/lib/srv.(*HostSudoersProvisioningBackend).CheckSudoers
Dec 20 20:04:35 __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/usermgmt_linux.go:141 github.com/gravitational/teleport/lib/srv.(*HostSudoersProvisioningBackend).WriteSudoersFile
Dec 20 20:04:35 __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/usermgmt.go:205 github.com/gravitational/teleport/lib/srv.(*HostSudoersManagement).WriteSudoers
Dec 20 20:04:35 __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/sess.go:248 github.com/gravitational/teleport/lib/srv.(*SessionRegistry).TryWriteSudoersFile
Dec 20 20:04:35  __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1709 github.com/gravitational/teleport/lib/srv/regular.(*Server).dispatch
Dec 20 20:04:35 __REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1577 github.com/gravitational/teleport/lib/srv/regular.(*Server).handleSessionRequests
Dec 20 20:04:35__REDACTED__ teleport[17226]: github.com/gravitational/teleport/lib/srv/regular/sshserver.go:1310 github.com/gravitational/teleport/lib/srv/regular.(*Server).HandleNewChan.func1
Dec 20 20:04:35__REDACTED__teleport[17226]: runtime/asm_amd64.s:1650 runtime.goexit
Dec 20 20:04:35 __REDACTED__  teleport[17226]: User Message: parse error in stdin near line 2
Dec 20 20:04:35 __REDACTED__ teleport[17226]: visudo: invalid sudoers file] regular/sshserver.go:2094
Dec 20 20:04:35__REDACTED__ teleport[17226]: 2023-12-20T20:04:35Z [AUDIT]     INFO session.data addr.remote:__REDACTED__code:T2006I ei:2.147483646e+09 event:session.data login:__REDACTED__ namespace:default private_key_policy:none rx:5250 __REDACTED__ sid: time:2023-12-20T20:04:35.924Z tx:7168 uid:__REDACTED__ user:__REDACTED__ events/emitter.go:271`

checked file
`sh-4.2$ sudo visudo -c
/etc/sudoers: parsed OK
/etc/sudoers.d/90-cloud-init-users: parsed OK
/etc/sudoers.d/ssm-agent-users: parsed OK
....
sh-4.2$`

**Reply by webvictim** [reply]
Can you share your Teleport role?

**Reply by code-dave** [reply]
role:

> kind: role
> metadata:
>   name: __SAMEGROUPNAME__
> spec:
>   allow:
>     app_labels:
>       '*': '*'
>     host_sudoers:
>     - 'ALL = (root) NOPASSWD: /usr/bin/systemctl restart nginx.service'
>     - 'echo ''@wheel ALL=(ALL:ALL) NOPASSWD: ALL'' | sudo tee /etc/sudoers.d/wheel'
>     kubernetes_labels:
>       '*': '*'
>     kubernetes_resources:
>     - kind: pod
>       name: '*'
>       namespace: '*'
>       verbs:
>       - '*'
>     logins:
>     - '{{email.local(external.username)}}'
>     - admin
>     - ec2-user
>     node_labels:
>       '*': '*'
>   deny: {}
>   options:
>     cert_format: standard
>     create_db_user: false
>     create_desktop_user: false
>     create_host_user_mode: drop
>     desktop_clipboard: true
>     desktop_directory_sharing: true
>     enhanced_recording:
>     - command
>     - network
>     forward_agent: false
>     idp:
>       saml:
>         enabled: true
>     max_session_ttl: 24h0m0s
>     pin_source_ip: false
>     port_forwarding: true
>     record_session:
>       default: best_effort
>       desktop: true
>     ssh_file_copy: true
> version: v5

**Comment by CyberPingU** [comment]
```
kind: role
metadata:
  description: Server God Mode
  id: 1702474428877878754
  labels:
    teleport.internal/resource-type: preset
  name: SGM
spec:
  allow:
    app_labels:
      '*': '*'
    host_groups:
    - sudo
    - wheel
    host_sudoers:
    - '%sudo ALL=(ALL) NOPASSWD: ALL'
    logins:
    - '{{internal.logins}}'
    node_labels:
      '*': '*'
    rules:
    - resources:
      - event
      verbs:
      - list
      - read
    - resources:
      - session
      verbs:
      - read
      - list
    - resources:
      - instance
      verbs:
      - list
      - read
    - resources:
      - assistant
      verbs:
      - list
      - create
      - read
      - update
      - delete
      - use
  deny: {}
  options:
    cert_format: standard
    create_db_user: false
    create_desktop_user: false
    create_host_user_mode: drop
    desktop_clipboard: true
    desktop_directory_sharing: true
    enhanced_recording:
    - command
    - network
    forward_agent: true
    idp:
      saml:
        enabled: true
    max_session_ttl: 30h0m0s
    pin_source_ip: false
    port_forwarding: true
    record_session:
      desktop: true
    ssh_file_copy: true
version: v7
```

**Comment by yair-segal** [comment]
I am getting the same error, with the same example. Any update on this?

**Comment by code-dave** [comment]
Same. I'm still having this problem

**Comment by yair-segal** [comment]
I don't know if I fixed the underlying issue, but I managed to make it work, after I noticed this comment on the docs (in [this section](https://goteleport.com/docs/server-access/guides/host-user-creation/#step-13-configure-a-role)):

> When a Teleport user accesses an SSH Service instance, Teleport checks each of the user's roles that match the instance. If at least one role matches the instance but does not specify create_host_user_mode to be either keep, drop, or insecure-drop, automatic user creation will be disabled. Roles that do not match the server will not be checked.

I had 2 roles that matched the node. Once I removed the one that didn't have "create_host_user_mode" (for the sake of testing) -- I was able to ssh to the server, and the error message was gone.

Since, in my opinion, this has nothing to do with the "sudo: invalid sudoers file" error message, I can't be 100% my fix is related. But at least this allows me to move forward for now.
